### PR TITLE
1908 1909 1910 performance optimizations 1

### LINKED ISSUE #1908: Disable debug print preConfig evaulation when in production mode
**What Needs to be Done?**

- [ ] First do this
- [ ] Then do that.
- [ ] Also this other thing.

**Is your feature request related to a problem? Please describe.**

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #1909: Cache the MPI communicator in AM
**What Needs to be Done?**

- [ ] First do this
- [ ] Then do that.
- [ ] Also this other thing.

**Is your feature request related to a problem? Please describe.**

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #1910: Break out of tryProcessData in AM early
**Describe the bug**

**To Reproduce**
1. Example/test/snippet of code that fails
2. Compiler, platform, libraries
3. Run command: number of processors, threading options, etc.

**Expected behavior**

**Screenshots**

**Platform (please complete the following information):**
 - Supercomputer or desktop [e.g., Trinity, Sierra)
 - Compiler environment [e.g., clang-3.9, gcc8]
 - Libraries [e.g., mpich-2, openmpi)
 - Commit ID or release of VT [e.g., vt-1.0.0]

**Additional context**

### PR DESCRIPTION
**Pull request by stmcgovern**
Closes #1908.
Closes #1909.
Closes #1910.
First PR of multiple for fine-grained performance enhancements.

### CODE REVIEW

File: src/vt/configs/debug/debug_print.h [hunk 1]
```diff
@@ -91,6 +95,7 @@
     vt::config::CtxEnum::ctx,                                           \
     vt::config::ModeEnum::inmode                                        \
   >::apply(vt_debug_argument_option(cat) or force, __VA_ARGS__)
+#endif
```
Comment (PhilMiller) [inline src/vt/configs/debug/debug_print.h:98]: Though it's small, I think it's preferable to put this in a separate PR, for easier traceability.

File: src/vt/configs/debug/debug_print.h [hunk 2]
```diff
@@ -83,6 +83,10 @@
 #define vt_debug_argument_option(opt)                                   \
   ::vt::debug::preConfig()->vt_debug_ ## opt
 
+#if vt_check_enabled(production_build)
+#define vt_debug_print_impl(force, inconfig, inmode, cat, ctx, ...)     \
+  vt_force_use(__VA_ARGS__)
```
Comment (PhilMiller) [inline src/vt/configs/debug/debug_print.h:88]:
```suggestion
  do { if (false) vt_force_use(__VA_ARGS__); } while (false);
```

**Comment by stmcgovern** [comment]
Split into PR #1924, #1925, and #1926.
